pyClient/zeth/joinsplit.py

```diff
@@ -0,0 +1,723 @@
+from __future__ import annotations
+import zeth.constants as constants
+import zeth.errors as errors
+from zeth.utils import get_trusted_setup_dir, hex_extend_32bytes, \
+    hex_digest_to_binary_string, string_list_flatten, encode_single, \
+    encode_abi, encrypt, decrypt, get_public_key_from_bytes
+from zeth.prover_client import ProverClient
+from api.util_pb2 import ZethNote, JoinsplitInput, HexPointBaseGroup1Affine, \
+    HexPointBaseGroup2Affine
+from nacl.public import PrivateKey, PublicKey  # type: ignore
+import nacl.utils  # type: ignore
+import api.prover_pb2 as prover_pb2
+
+import os
+import json
+import sys
+from Crypto import Random
+from hashlib import blake2s, sha256
+from py_ecc import bn128 as ec
+from typing import Tuple, Dict, List, Iterable, Union, Any
+
+
+FQ = ec.FQ
+G1 = Tuple[ec.FQ, ec.FQ]
+
+
+class ApkAskPair:
+    def __init__(self, a_sk: str, a_pk: str):
+        self.a_pk = a_pk
+        self.a_sk = a_sk
+
+
+SK = Tuple[int, int]
+VK = Tuple[G1, G1]
+
+
+class VkSkPair:
+    def __init__(self, x: int, y: int, x_g1: G1, y_g1: G1):
+        self.vk = (x_g1, y_g1)
+        self.sk = (x, y)
+
+
+# Dictionary representing a VerificationKey from any supported snark
+GenericVerificationKey = Dict[str, Any]
+
+
+# Dictionary representing a Proof from any supported snark
+GenericProof = Dict[str, Any]
+
+
+# class ZkSnarkProvider(object):
+#     def parseVerificationKey(
+#             vkObj: prover_pb2.VerificationKey) -> GenericVerificationKey:
+#         pass
+
+#     def writeVerificationKey(
+#         vkObj: prover_pb2.VerificationKey) -> None:
+#         pass
+
+#     def parseProof(
+#         proofObj: prover_pb2.ExtendedProof) -> GenericProof:
+#         pass
+
+#     def deploy_verifier_contract(
+#             vk: prover_pb2.VerificationKey,
+#             deployer_address,
+#     ) -> str:
+#         """
+#         Return address of verifier contract
+#         """
+#         pass
+
+#     def deploy mixer_contract() -> Tuple[Instance]:
+#         pass
+
+#     def encode_and_hash_for_contract(
+#             proof: GenericProof) ->
+#     {
+#         proof_digest: bytes,
+#         inputs_digest: bytes,
+#         contract_input: ContractProof
+#         }
+
+#     def call_mixer(
+#             mixer_instance: Any,
+#             ...
+#             proof: ContractProof,
+#             ...
+#             other params from mix_groth16...)
+
+#     #   - hash_proof, hash_inputs for signing
+
+
+class NoteRandomness:
+    def __init__(self, rho: str, trap_r: str):
+        self.rho = rho
+        self.trap_r = trap_r
+
+    @staticmethod
+    def new() -> NoteRandomness:
+        """
+        Compute the note randomness: the trapdoor trapR and rho. Starting the
+        Non-Malleability update, rho is computed from phi (see above), the rho
+        generated in this function is thus obsolete except for dummy input
+        notes.
```

Comment:
I'm not sure it is a good idea to stick to this function since rho needs to be derived in another way (from phi as mentioned in the comment)

Reply:
Yeah, I was wondering what to do about this.
I've removed the `NoteRandomness` and now just have a `trap_r_randomness()` that returns a random `r`.
All the obsolete stuff is moved to `mock.py` since that's the only place it is used.